11-payment-encoding.md

```diff
@@ -0,0 +1,345 @@
+   `multiplier`.
+
+The following `multiplier` letters are defined:
+
+* `m` (milli): multiply by 0.001
+* `u` (micro): multiply by 0.000001
+* `n` (nano): multiply by 0.000000001
+* `p` (pico): multiply by 0.000000000001
+
+## Requirements
+
+A writer MUST include `amount` if payments will be refused if less
+than than that.  A writer MUST encode `amount` as a positive decimal
+integer with no leading zeroes, SHOULD use the shortest representation
+possible.
+
+A reader MUST fail if it does not understand the `prefix`.  A reader
+SHOULD fail if `amount` contains a non-digit, or is followed by
+anything except a `multiplier` in the table above.
+
+A reader SHOULD indicate if amount is unspecified, otherwise it MUST
+multiply `amount` by the `multiplier` value (if any) to derive the
+amount required for payment.
+
+## Rationale
+
+The `amount` is encoded into the human readable part, as it's fairly
+readable and a useful indicator of how much is being requested.
+
+Donation addresses often don't have an associated amount, so `amount`
+is optional in that case: usually a minimum payment is required for
+whatever is being offered in return.
+
+# Data Part
+
+The data part consists of multiple sections:
+
+1. `version`: 0 (5 bits)
+1. `timestamp`: seconds-since-1970 (35 bits, big-endian)
+1. `payment_hash`: (256 bits + 4 trailing zero bits)
+1. Zero or more tagged parts.
+1. `signature`: bitcoin-style signature of above. (520 bits)
+
+## Requirements
+
+A writer MUST set `version` to 0, MUST set `timestamp` to the time to
+the number of seconds since Midnight 1 January 1970, UTC in
+big-endian, and MUST set `payment_hash` to the SHA-2 256-bit hash of
+the `payment_preimage` which will be given in return for payment,
+followed by four zero bits.  A writer MUST set `signature` to a valid
+512-bit secp256k1 signature of the double SHA2 256-bit hash of the
+Human Readable Part concatenated with a bytes for each 5 bits of the
+Data Part, with a trailing byte containing the recovery ID (0, 1, 2 or
+3).
+
+## Rationale
+
+The 5-bit `version` fits neatly into base32 encoding (0 == `q`), and
+allows for future incompatible upgrades.  `signature` covers the
+pre-base32 encoded data for simplicity, especially since the data part
+may not end on an 8-bit boundary.  It also allows public key recovery,
+so the identity of the payee node is implied.
+
+## Tagged Fields
+
+Each Tagged Field is of format:
+
+1. `type` (5 bits)
+1. `length` (10 bits, big-endian)
+1. `data` (length x 5 bits)
+
+Currently defined Tagged Fields are:
+
+* `d` (13): short description of purpose of payment (ASCII),  e.g. '1 cup of coffee'
+* `h` (23): 256-bit description of purpose of payment (SHA256).  This is used to commit to an associated description which is too long to fit, such as may be contained in a web page.
+* `x` (6): `expiry` time in seconds (big-endian). Default is 3600 (1 hour) if not specified.
+* `f` (9): fallback on-chain address.  For bitcoin, this is 5 bits of
```

Comment:
Maybe the invoice format should be compatible with BIP 21, so existing clients can use the fallback?